# Atividade em Dupla 1

## Descrição dâ base de dâdos escolhida:

### O que signigica cada linha de cada arquivo?
$R:$ 

### Quais são os principais atributos (colunas) e seus tipos?
$R:$ 

## Descrição da base de dados escolhida:

### Teve dificuldades para preparar o dataset para uso? (nomes ruins para colunas, arquivos dos datasets não possuíam boa organização?)
$R:$ 

### Aplicou algum pré-processamento nos dados?
$R:$ Sim, para a Pergunta 1 da Seção de Análise Exploratória 




## Análise Exploratória

In [5]:
# Carregando bibliotecas
import pandas as pd
import os 

### Loading MovieLens DFs

In [101]:
# Getting current working directory
main_path = os.getcwd()

# Building file paths 
path_movielens_movies = main_path + '/movielens/movies.csv'
path_movielens_ratings = main_path + '/movielens/ratings.csv'
path_movielens_tags = main_path + '/movielens/tags.csv'

# Loading files into DFs
df_movies = pd.read_csv(path_movielens_movies)
df_ratings = pd.read_csv(path_movielens_ratings)
df_tags = pd.read_csv(path_movielens_tags)

### Checking DFs

In [118]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


In [117]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [116]:
df_tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   userId     3683 non-null   int64 
 1   movieId    3683 non-null   int64 
 2   tag        3683 non-null   object
 3   timestamp  3683 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 115.2+ KB


### Pergunta 1 – Existe alguma correlação entre o ano de lançamento do filme e sua avaliação média (média de ratings)?

In [18]:
import re

In [180]:
# Testing regex logic to get year 

def get_year_from_raw_title(raw_title):
    regex_exp = re.compile('\((....)\)')
    try:
        year = int(re.sub('[()]', '', re.search(regex_expression, raw_title).group()))
    except: 
        year = None
    return year

movie_list_for_test = list(df_movies.sample(n=5).title)

for movie in movie_list_for_test:
    print(f" Raw Title: {movie}\nMovie Year: {get_year_from_raw_title(movie)}\n")

 Raw Title: Think Like a Man Too (2014)
Movie Year: 2014

 Raw Title: Scarlet Letter, The (1995)
Movie Year: 1995

 Raw Title: Three Musketeers, The (1993)
Movie Year: 1993

 Raw Title: TMNT (Teenage Mutant Ninja Turtles) (2007)
Movie Year: 2007

 Raw Title: Prom Night IV: Deliver Us From Evil (1992)
Movie Year: 1992



In [181]:
# Testing regex logic to remove year from title 

def remove_year_from_raw_title(raw_title):
    regex_exp = re.compile('\((....)\)')
    try:
        title = re.sub(regex_expression, '', raw_title)
    except: 
        return raw_title
    return title

movie_list_for_test = list(df_movies.sample(n=5).title)

for movie in movie_list_for_test:
    print(f"Old Title: {movie}\nNew Title: {remove_year_from_raw_title(movie)}\n")

Old Title: Party Girl (1995)
New Title: Party Girl 

Old Title: To Do List, The (2013)
New Title: To Do List, The 

Old Title: Hollywood Shuffle (1987)
New Title: Hollywood Shuffle 

Old Title: How to Be Single (2016)
New Title: How to Be Single 

Old Title: Shenandoah (1965)
New Title: Shenandoah 



In [182]:
# Pre Procesing Data
# Creating 'year' column and adjusting the title - Movie DF

# Extracting Years
years = []
for title in df_movies['title']:
    years.append(get_year_from_raw_title(title))

# Removing year from title
titles = []
for title in df_movies['title']:
    titles.append(remove_year_from_raw_title(title))


# Creating new DF with title normalized and 'year' column
df_movies_title_normalized_year_column = df_movies[['movieId','title']]
df_movies_title_normalized_year_column['year'] = years
df_movies_title_normalized_year_column['title'] = titles
df_movies_title_normalized_year_column.head()

movieId                         title    year
0        1                    Toy Story   1995.0
1        2                      Jumanji   1995.0
2        3             Grumpier Old Men   1995.0
3        4            Waiting to Exhale   1995.0
4        5  Father of the Bride Part II   1995.0

In [190]:
# AVG Rate

df_ratings_groupby_movieId_with_mean_rating = df_ratings.groupby('movieId').mean()[['rating']].round(2)
df_ratings_groupby_movieId_with_mean_rating.head()

rating
movieId        
1          3.92
2          3.43
3          3.26
4          2.36
5          3.07

In [191]:
# Merging rating DF with Title and Year DF

df_movie_ratings_merged = pd.merge(df_ratings_groupby_movieId_with_mean_rating, df_movies_title_normalized_year_column, how = 'inner', on = 'movieId')

correlation = df_movie_ratings_merged['rating'].corr(df_movie_ratings_merged['year'])
print(f'Correlation Coefficient: {round(correlation, 2)}')

Correlation Coefficient: -0.11


$R:$ Correlation Coefficient: -0.11 - $Falta a Explicação$

### Pergunta 2 – Qual o grupo mais disperso, o de usuários que avaliaram ou o de usuários que rotularam filmes?

### Pergunta 3 – Qual a categoria rotulada mais frequente?

### Pergunta 4 – Como estão distribuídas as avaliações ao longo do tempo?